src/EVM/Expr.hs

```diff
@@ -628,6 +633,32 @@ getAddr (GVar _) = error "cannot determine addr of a GVar"
 
 -- ** Whole Expression Simplification ** -----------------------------------------------------------
 
+-- Calls prop simplification for the appropriate types of Expr
+applyToProps :: ([Prop] -> [Prop]) -> Expr b -> Expr b
+applyToProps f expr = case expr of
+  Failure a b c -> Failure (f a) b c
+  Partial a b c -> Partial (f a) b c
+  Success a b c d -> Success (f a) b c d
+  ITE (Lit 0) _ c -> applyToProps f c
+  ITE (Lit _) b _ -> applyToProps f b
+  ITE a b c-> ITE a (applyToProps f b) (applyToProps f c)
+  a -> a
+
+remRedundantProps :: [Prop] -> [Prop]
+remRedundantProps p = collapseFalse . filter (\x -> x /= PBool True) . nubOrd $ p
+  where
+    collapseFalse :: [Prop] -> [Prop]
+    collapseFalse ps = if isJust $ find (== PBool False) ps then [PBool False] else ps
+
+simplifyPropExpr :: Expr a -> Expr a
```

Comment:
```suggestion
simplifyProps :: Expr End -> Expr End
```

I think we can restrict the type here?

Reply:
actually I think this can just be inlined into the existing expression simplifier by calling `remRedundantProps . map evalProp . flattenProps` on `Success` / `Failure` / `Partial`? Then we can get rid of `applyToProps` and `simplifyPropExpr`?

Reply:
I made this a bit cleaner in 306f7087118856193460d765122977666a5f5fa5 Let me know what you think. However, it's not inlined, because I don't understand how to do that. Let's discuss in person :)